packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -38,27 +90,177 @@ contract FundingRateStore is FundingRateStoreInterface, Testable {
 
         // TODO: bond logic.
 
         require(proposalState != ProposalState.Pending, "Existing proposal still pending.");
 
         if (proposalState == ProposalState.Expired) {
+            // Publish expired rate, reward proposer.
             fundingRateRecord.rate = fundingRateRecord.proposal.rate;
+
+            // TODO: Reward = proposal bond
+            collateralCurrency.safeTransfer(
+                fundingRateRecord.proposal.proposer,
+                fundingRateRecord.proposal.finalFee.rawValue
+            );
+
+            emit PublishedRate(
+                identifier,
+                fundingRateRecord.rate.rawValue,
+                fundingRateRecord.proposal.time,
+                fundingRateRecord.proposal.proposer
+            );
+        }
+
+        // Compute final fee at time of proposal.
+        FixedPoint.Unsigned memory finalFeeBond = _computeFinalFees();
+
+        uint256 currentTime = getCurrentTime();
+        fundingRateRecord.proposal = Proposal({
+            rate: rate,
+            time: currentTime,
+            proposer: msg.sender,
+            finalFee: finalFeeBond,
+            disputer: address(0x0)
+        });
+        emit ProposedRate(identifier, rate.rawValue, currentTime, msg.sender);
+
+        // Pull final fee bond from proposer.
+        collateralCurrency.safeTransferFrom(msg.sender, address(this), finalFeeBond.rawValue);
+    }
+
+    /**
+     * @notice Dispute a pending funding rate. This will delete the pending funding rate, meaning that a
+     * proposer can now proposer another rate with a fresh liveness.
+     * @dev This will revert if there is no pending funding rate for the identifier.
+     * @dev Caller must approve this this contract to spend `finalFeeBond` amount of collateral, which they can
+     * receive back if their dispute is successful.
+     */
+    function dispute(bytes32 identifier) external {
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(identifier);
+        ProposalState proposalState = _getProposalState(fundingRateRecord.proposal);
+
+        require(proposalState == ProposalState.Pending, "Existing proposal not pending.");
+
+        // TODO: Bond logic.
+
+        // Pull final fee bond from disputer and pay store.
+        _payFinalFees(msg.sender, fundingRateRecord.proposal.finalFee);
+
+        // Send price request
+        _requestOraclePrice(identifier, fundingRateRecord.proposal.time);
+
+        emit DisputedRate(
+            identifier,
+            fundingRateRecord.proposal.rate.rawValue,
+            fundingRateRecord.proposal.time,
+            fundingRateRecord.proposal.proposer,
+            msg.sender
+        );
+
+        // Delete pending proposal and copy into dispute records.
+        fundingRateRecord.proposal.disputer = msg.sender;
+        fundingRateDisputes[identifier][fundingRateRecord.proposal.time] = fundingRateRecord;
```

Comment:
Might be a weird question, but what if someone makes a proposal, disputes it, and then makes a second proposal and disputes it in the same txn? Would that break anything?

Reply:
Proposer would post a final fee bond, disputer would post a final fee bond. 

Second proposal would post a final fee bond. 

Now assume that the second proposal expires before the dispute settles, then the second proposal will get published. Once the dispute settles it will again publish a new rate. 

I think we talked about this and I think i haven't implemented it to the spec, but should we have the settled dispute publish a new rate IFF there have been no published proposals since the dispute started?

Reply:
Yeah, I see what you're saying, and that's definitely something worth discussing (I think we're covering a similar thing in the other thread). I'm talking about something a little lower level: won't the second dispute overwrite the first dispute in this data structure (same keys)? Would that cause any bad things to happen? Would the worst case just be that the person doing this would get their money locked up indefinitely? Are there any weird (positive EV) attack vectors that this overwriting would allow?

Reply:
Ah. The first dispute deletes the pending proposal. So the second dispute should not be possible. Do you see something diff?

Reply:
Yeah, the situation I'm talking about is the following steps in a single transaction:
1. Create proposal.
2. Dispute proposal.
3. Create proposal.
4. Dispute proposal.

Since they all have the same timestamp, would the second dispute overwrite the 1st? What are the consequences, if any, of that happening?

Reply:
All in the same block I'm assuming? I think you're right the second dispute would overwrite the first dispute, and this one user would pay 4 final fees to earn back 1. The final fees from (1) and (2) would effectively be lost.

This makes me think of another bad outcome:
- Block 1:
    - User 1 creates a proposal
- Block 2:
    - User 2 disputes User 1's proposal
    - User 2 creates a proposal
    - User 2 disputes User 2's proposal

User 2 pays 3 final bonds to win back 1, but User 1 loses their funds. This would be uneconomical, but not good. WDYT?

Reply:
I dont see any issue with this as the disputer always ends up losing more final fees than what they get back, making this strictly uneconomical. It does spam the DVM but multiple queries of the price at the same timestamp for the same identifier will not be separate votes so the DVM does not care too much.

Reply:
interesting... this is really a question of if you should uniquely identify each dispute vs using timestamp as id.
perhaps you can require that `fundingRateDisputes[identifier][getCurrentTime()]` doesnt exist when adding a new proposal

Reply:
@daywiss That's a good idea. I agree with you the question is do we care enough to prevent this edge case? This one liner of `require(!fundingRateDisputes[identifier][getCurrentTime()])` in the proposal method would only serve to protect against single transactions containing multiple create+dispute proposal transactions.

Reply:
yea im not sure if it matters or not!

Reply:
It seems bad to have any scenario where storage is unwittingly overwritten and collateral gets locked up, although I agree that it doesn't seem to cause issues for anyone other than the caller.

Any reason not to disallow it out of principle/conservatism (maybe there's an attack vector we haven't thought of, like griefing a legit proposer?) the way David has suggested?

Reply:
Nope, I'll prevent it

Reply:
Using @daywiss 's recommended logic I've implemented this check along with a unit test in this [commit](https://github.com/UMAprotocol/protocol/pull/2147/commits/e29b8284d014d42e781bcfd268cd9b53a11e10e2). I think I've described and tested the issue here but not 100% sure.

Also @mrice32 I think your attack vector described below can be reduced to only steps 1,2,3 since the main risk is that the original proposer's bond is lost, and the second dispute doesn't really add anything to the "attack"

> 1. Create proposal.
> 2. Dispute proposal.
> 3. Create proposal.
> 4. Dispute proposal.

Reply:
@nicholaspai But in that case, wouldn't that original dispute end up getting resolved and the original proposal getting their bond back (if they were right)?

Either way, SGTM!

Reply:
Yeah isn't that by design?

Reply:
Yep, all I'm saying is that you don't get the overwrite if you remove step 4, so I think no money is lost and everything works as expected. I think you need step 4 to have the contract do something outside of the spec.

Reply:
Ah yes you're right

Reply:
So do you agree that by blocking step 3, which is what David's suggestion does and what I implemented, prevents step 4 from being an issue?

Reply:
I don't really see a way to block step 4 explicitly unless we create some sort of way to store multiple disputes per <proposal-time, identifier>, which I think is just a bit more complex.

Reply:
As discussed offline, 100% agree with your conclusion here @nicholaspai. Blocking 3 seems to be the simplest solution.